server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoTileUtils.java

```diff
@@ -0,0 +1,195 @@
+    }
+
+    /**
+     * Encode lon/lat to the geotile based long format.
+     * The resulting hash contains interleaved tile X and Y coordinates.
+     * The precision itself is also encoded as a few high bits.
+     */
+    static long longEncode(double longitude, double latitude, int precision) {
+        // Mathematics for this code was adapted from https://wiki.openstreetmap.org/wiki/Slippy_map_tilenames#Java
+
+        // Number of tiles for the current zoom level along the X and Y axis
+        final int tiles = 1 << checkPrecisionRange(precision);
+
+        int xTile = (int) Math.floor((normalizeLon(longitude) + 180) / 360 * tiles);
+
+        double latSin = Math.sin(Math.toRadians(normalizeLat(latitude)));
+        int yTile = (int) Math.floor((0.5 - (Math.log((1 + latSin) / (1 - latSin)) / (4 * Math.PI))) * tiles);
+
+        // Edge values may generate invalid values, and need to be clipped.
+        // For example, polar regions (above/below lat 85.05112878) get normalized.
+        if (xTile < 0) {
+            xTile = 0;
+        }
+        if (xTile >= tiles) {
+            xTile = tiles - 1;
+        }
+        if (yTile < 0) {
+            yTile = 0;
+        }
+        if (yTile >= tiles) {
+            yTile = tiles - 1;
+        }
+
+        // Zoom value is placed in front of all the bits used for the geotile
+        // e.g. when max zoom is 29, the largest index would use 58 bits (57th..0th),
+        // leaving 5 bits unused for zoom. See MAX_ZOOM comment above.
+        return ((long) precision << ZOOM_SHIFT) | ((long) xTile << MAX_ZOOM) | ((long) yTile);
+    }
+
+    /**
+     * Parse geotile hash as zoom, x, y integers.
+     */
+    private static int[] parseHash(long hash) {
+        final int zoom = (int) (hash >>> ZOOM_SHIFT);
+        int xTile = (int) ((hash >>> MAX_ZOOM) & X_Y_VALUE_MASK);
+        int yTile = (int) (hash & X_Y_VALUE_MASK);
+        return new int[]{zoom, xTile, yTile};
+    }
+
+    /**
+     * Encode to a geotile string from the geotile based long format
+     */
+    static String stringEncode(long hash) {
+        int[] res = parseHash(hash);
+        validateZXY(res[0], res[1], res[2]);
+        return "" + res[0] + "/" + res[1] + "/" + res[2];
+    }
+
+    /**
+     * Decode long hash as a GeoPoint (center of the tile)
+     */
+    static GeoPoint hashToGeoPoint(long hash) {
+        int[] res = parseHash(hash);
+        return zxyToGeoPoint(res[0], res[1], res[2]);
+    }
+
+    /**
+     * Decode a string bucket key in "zoom/x/y" format to a GeoPoint (center of the tile)
+     */
+    static GeoPoint keyToGeoPoint(String hashAsString) {
+        Throwable cause = null;
+        try {
+            final String[] parts = hashAsString.split("/", 4);
+            if (parts.length == 3) {
+                return zxyToGeoPoint(Integer.parseInt(parts[0]), Integer.parseInt(parts[1]), Integer.parseInt(parts[2]));
+            }
+        } catch (IllegalArgumentException e) {
```

Comment:
Just curious about this arrangement... what's the reason for just catching `IllegalArgumentException` and then rethrowing it?  Should we just catch any `Exception`?  I'm assuming this is here to make the exception message a little nicer?